### ISSUE TITLE
Optitrack Connection Issue - Ubuntu 20.04

### ISSUE THREAD

**Issue by The-SS** [whoenig/NatNetSDKCrossplatform#3]
Hi Wolfgang, 

When I use this package to get Optitrack raw data on a machine running Ubuntu 20.04, I get the following output
```
$ ./natnettest 192.168.1.133
1 264
NatNetVersion: 0.0.0.0
ServerVersion: 2.10.0.0
```
I can ping 192.168.1.133 and I get Optitrack rigid bodies using vrpn_client_ros, but when I try to get the point cloud, I am out of luck.

I tried using an older machine running Ubuntu 16.04 and actually got the raw point cloud data.

I have the wired network configuration on both machines setup basically the same and just moving the OptiTrack ethernet cable between the two machines.
Do you have any ideas about why the Ubuntu 20.04 machine is not receiving the data? 

Thank you.
-Sleiman

**Comment by whoenig** [comment]
I use this with Ubuntu 20.04 and Motive 2.3 and 3.0 without issues.

One thing you could try is the official SDK (it does have a Linux version now, although this is not open-source). If the official one works, this would indicate that there is a bug in this repo, otherwise it might be a configuration problem in Motive or the network configuration (I don't recall that I changed anything on my Ubuntu).

**Comment by The-SS** [comment]
I downloaded the [NatNet SDK 4.0.0 ](https://optitrack.com/support/downloads/developer-tools.html) for Linux, compiled it per the instructions and ran `./SampleClient 192.168.1.133` (where 192.168.1.133 is the IP address of the Windows machine with Motive 1.10). I received raw point cloud data over to my Ubuntu 20.04 machine. Here is a sample:
```
FrameID : 2403470
Timestamp : 20028.92
Software latency : -nan milliseconds
Transit latency : -inf milliseconds
Timecode : 00:00:01:36.0
Rigid Bodies [Count=0]
Skeletons [Count=0]
Markers [Count=7]
Labeled Marker [ModelID=0, MarkerID=6557] [size=0.02] [pos=0.10,-0.12,0.11]
Labeled Marker [ModelID=0, MarkerID=6581] [size=0.02] [pos=0.32,-0.08,0.11]
Labeled Marker [ModelID=0, MarkerID=6582] [size=0.02] [pos=0.12,0.02,0.14]
Labeled Marker [ModelID=0, MarkerID=6583] [size=0.02] [pos=0.29,0.08,0.11]
Labeled Marker [ModelID=0, MarkerID=6586] [size=0.02] [pos=0.07,0.04,0.12]
Labeled Marker [ModelID=0, MarkerID=6587] [size=0.02] [pos=0.19,-0.08,0.13]
Labeled Marker [ModelID=0, MarkerID=6591] [size=0.02] [pos=0.07,-0.03,0.12]
Force Plate [Count=0]
Device [Count=0]
```
Having confirmed that, I tried this package again, but `./natnettest 192.168.1.133` returns
```
NatNetVersion: 0.0.0.0
ServerVersion: 2.10.0.0
```

I also tried your [other libmotioncapture package](https://github.com/IMRCLab/libmotioncapture). When I run `./motioncapture_example optitrack 192.168.1.133` from `libmotioncapture/build`, I get:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Motive does not use multicast. Please update your streaming settings!
Aborted (core dumped)
```
(my Motive settings have Multicast enabled with command port 1510 and data port 1511).

It seems that I can (through the official SDK or VRPN) receive data but this package is having some issues. Any ideas what I can check/edit to make this work?

**Comment by whoenig** [comment]
I have heard from multiple people that there are issues with Motive 1.x. This is tricky to debug, since I don't have that version. What is your use case? One option would be to add support for the pre-built SDK in libmotioncapture. The other option would be to debug/analyze the packet stream that you get from Motive 1.x. The third option would be to upgrade to Motive 2.x or 3.0, both of which are tested and I have access to.

**Comment by The-SS** [comment]
I'm setting up Crazyswarm on this new machine. Since `mocap_helper.launch` didn't show any raw data, I tried checking with this (as you've recommended in some other issues/discussions). Since it seemed that the issue was not in Crazyswarm but rather on the libmotioncapture side, I brought up this issue here.

**Comment by The-SS** [comment]
I upgraded to Motive 2.0.1 (cannot use anything newer for now). The script responds differently now.
On the new machine with Ubuntu 20.04, `/natnettest 192.168.1.133` returns:
```
NatNetVersion: 88.208.50.0
ServerVersion: 3.0.0.0
```
then nothing happens.
On the new machine with Ubuntu 16.04, I get the marker data after that.

NatNet SDK 4.0.0 returns the same data on both machines.

**Comment by whoenig** [comment]
what about https://github.com/IMRCLab/libmotioncapture in the case of Motive 2.0.1?

**Comment by The-SS** [comment]
That returns 
```
supportsRigidBodyTracking: 1
supportsLatencyEstimate: 1
supportsPointCloud: 1
supportsTimeStamp: 1
```
then nothing happens.
(I only ran this on the Ubuntu 20.04 machine).

**Comment by whoenig** [comment]
I think the best short-term solution will be to add a new backend to libmotioncapture that links to NatNetSDK 4.0, see https://github.com/IMRCLab/libmotioncapture/issues/1. PRs are welcome - I am not sure when exactly I'd get to implement this myself (probably about 2h of work in total).

**Comment by The-SS** [comment]
Which version of Boost are you using?

**Comment by whoenig** [comment]
the default of 20.04: 1.71

**Comment by The-SS** [comment]
Summary so far: this package works with Motive 1.10.4 or Motive 2.0.1 with Ubuntu 16.04 (Boost version 1.58) but not with Ubuntu 20.04 (Boost version 1.71).

**Comment by whoenig** [comment]
I doubt it's because of boost. I added support for the official SDK.

* Here, you can update this repo and try it (see updated readme for instructions to run).
* In https://github.com/IMRCLab/libmotioncapture/blob/main/CMakeLists.txt#L8 you need to enable this new feature. Then, you use as type "optitrack_closed_source" rather than "optitrack"
* To use it with the Crazyswarm, simply update the libmotioncapture submodule to the latest version and enable https://github.com/IMRCLab/libmotioncapture/blob/main/CMakeLists.txt#L8

Please report back if this works for you.

**Comment by The-SS** [comment]
Thanks for adding that so quickly!

I have several updates about this issue.
- We were able to upgrade to Motive 2.3. Despite the update, we're still stuck with the same issue. VRPN and NatNetSDK work fine, but the open-source package fails.
We did a fresh Ubuntu 20.04 install and while installing ROS Noetic, it crashed due to an error with libboost1.71. The crash was due to the user not having root privileges which caused a specific file to not be created. Rerunning `sudo apt install ros-noetic-desktop-full` a second time completes the installation without any errors. I suspect that this error that happened has something to do with the package failing. 
We're going to do a new fresh OS install and install ROS with root privileges. I'm hoping that this resolves the issue and we could use the open-source package directly. I'll let you know how things go.
- Regarding the updates to `libmotioncapture` and `NatNetSDKCrossplatform`, the added support for the closed source SDK worked! 
The only notes I have about that are: 
    - running `make` to compile `libmotioncapture` is missing [from the documentation](https://github.com/IMRCLab/libmotioncapture/blob/13d14f85d29171426a2b663cf71cb6262c6f7760/README.md?plain=1#L44)
    - after changing [ENABLE_OPTITRACK_CLOSED_SOURCE ](https://github.com/IMRCLab/libmotioncapture/blob/13d14f85d29171426a2b663cf71cb6262c6f7760/CMakeLists.txt#L8) to ON, users should remove everything in `build` before recompiling.
- As for the integration with Crazyswarm, this was not straightforward. I pulled changes for both `NatNetSDKCrossplatform` and `libmotioncapture`, and even built them directly, but `catkin_make` was crashing because it wasn't finding some NatNet libraries. We're doing a fresh OS install now, so I didn't have much time to debug this. 

TL;DR: The closed-source SDK support worked but integration with Crazyswarm was not tested. This seems to be an issue with our setup and not with the package itself, so older versions of Motive might still work with the open-source solution.

I'll close the issue here. If we get the Crazyswarm integration done, I'll let you know. If we have problems with that integration, I'll open an issue on Crazyswarm.

Thanks for your help!

Update: even with the fresh Ubuntu 20.04 install and installing ROS with no errors, the package is still not receiving data from Motive 2.3 using the open-source implementation. It does however receive data using the closed-source SDK.
